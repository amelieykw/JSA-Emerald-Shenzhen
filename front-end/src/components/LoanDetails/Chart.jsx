import React, { useEffect, useState } from 'react'
import { VictoryChart, VictoryLine, VictoryScatter } from 'victory-native'
import { View } from 'react-native'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'

const Chart = ({ loanHistory }) => {
  let mounted = true
  const [dataPair, setDataPair] = useState([])

  useEffect(() => {
    if (mounted && loanHistory) {
      let isOneYear = true
      const year = parseInt(loanHistory[0].date, 10)
      loanHistory.forEach(histItem => {
        if (year !== parseInt(histItem.date, 10)) {
          isOneYear = false
        }
      })
      if (isOneYear) {
        const dayDataPair = loanHistory.map(histItem => ({
          x: `${histItem.date.split('-')[1]}-${histItem.date.split('-')[2]}`,
          y: histItem.remaining
        }))
        setDataPair(dayDataPair)
      } else {
        const yearDataPair = []
        loanHistory.forEach(histItem => {
          if (yearDataPair.length === 0) {
            yearDataPair.push({
              x: histItem.date,
              y: histItem.remaining
            })
          } else if (
            parseInt(histItem.date, 10) !== parseInt(yearDataPair[yearDataPair.length - 1].x, 10)
          ) {
            yearDataPair.push({
              x: histItem.date,
              y: histItem.remaining
            })
          }
        })
        setDataPair(yearDataPair)
      }
    }
    return () => {
      mounted = false
    }
  }, [loanHistory])

  return dataPair.length > 0 ? (
    <View>
      <VictoryChart height={250}>
        <VictoryLine
          style={{ data: { stroke: '#FB7348' }, parent: { border: '1px solid #ccc' } }}
          data={dataPair}
        />
        <VictoryScatter data={dataPair} size={5} style={{ data: { fill: '#FB7348' } }} />
      </VictoryChart>
    </View>
  ) : null
}

Chart.propTypes = {
  loanHistory: PropTypes.arrayOf(PropTypes.any)
}

Chart.defaultProps = {
  loanHistory: undefined
}

const mapStateToProps = state => {
  return {
    loanHistory: state.loanList.find(loanItem => loanItem.id === state.detailLoanId).history
  }
}

export default connect(mapStateToProps, null)(Chart)
