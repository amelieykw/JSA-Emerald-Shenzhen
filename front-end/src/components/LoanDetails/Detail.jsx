import React, { useContext } from 'react'
import { Text, Grid, Col, Icon } from 'native-base'
import { TouchableOpacity, View } from 'react-native'
import { NavigationContext } from 'react-navigation'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'

import style from './DetailStyle'

const Detail = ({ loanList, id }) => {
  const navigation = useContext(NavigationContext)
  const { type, remaining } = loanList.find(loan => loan.id === id)

  const handlePress = () => {
    navigation.navigate('PayNow')
  }

  return (
    <Grid style={style.body}>
      <Col style={style.col}>
        <Text style={style.id}>{`ID - ${id}`}</Text>
        <Text style={style.name}>{type.name}</Text>
        <Text>{`Current Interest rate: ${type.interest}%`}</Text>
      </Col>
      <Col style={style.col}>
        <View>
          <Text style={style.remaining}>{`$ ${remaining}`}</Text>
          <Text style={style.remainingText}>remaining</Text>
        </View>
        <TouchableOpacity style={style.button} onPress={handlePress}>
          <View style={style.circle}>
            <Icon type="FontAwesome" name="dollar" style={style.dollar} />
          </View>
          <Text>PAY NOW</Text>
        </TouchableOpacity>
      </Col>
    </Grid>
  )
}

const mapStateToProps = state => {
  return { loanList: state.loanList, id: state.detailLoanId }
}

Detail.propTypes = {
  loanList: PropTypes.arrayOf(PropTypes.any).isRequired,
  id: PropTypes.number.isRequired
}

export default connect(mapStateToProps, null)(Detail)
