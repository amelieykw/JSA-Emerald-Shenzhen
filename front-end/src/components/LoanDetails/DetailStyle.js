import { StyleSheet } from 'react-native'

const style = StyleSheet.create({
  body: {
    backgroundColor: '#E3F2FD',
    height: 120,
    padding: 15
  },
  col: {
    justifyContent: 'space-between'
  },
  id: {
    fontSize: 18,
    color: '#7A7A7A'
  },
  name: {
    fontSize: 32,
    color: '#5A5C60'
  },
  remaining: {
    fontSize: 36,
    color: '#FC815E',
    alignSelf: 'flex-end'
  },
  remainingText: {
    paddingRight: 5,
    fontSize: 12,
    alignSelf: 'flex-end',
    color: '#7A7A7A'
  },
  button: {
    flexDirection: 'row',
    justifyContent: 'flex-end',
    alignItems: 'center'
  },
  circle: {
    backgroundColor: 'black',
    height: 21,
    width: 21,
    borderRadius: 20,
    alignItems: 'center',
    marginRight: 8
  },
  dollar: {
    color: 'white',
    fontSize: 20
  }
})

export default style
