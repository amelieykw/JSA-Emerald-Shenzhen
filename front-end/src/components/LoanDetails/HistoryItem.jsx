import React from 'react'
import { View } from 'react-native'
import { Text, Icon } from 'native-base'
import PropTypes from 'prop-types'

import style from './HistoryItemStyle'

const HistoryItem = ({ historyItem }) => {
  return (
    <View style={style.body}>
      <View style={[style.circle, historyItem.amount > 0 ? style.up : style.down]}>
        <Icon
          type="Feather"
          name={`trending-${historyItem.amount > 0 ? 'up' : 'down'}`}
          style={style.icon}
        />
      </View>
      <View>
        <Text>{historyItem.text}</Text>
        <Text style={style.date}>{historyItem.date}</Text>
      </View>
    </View>
  )
}

HistoryItem.propTypes = {
  historyItem: PropTypes.shape({
    amount: PropTypes.number,
    text: PropTypes.string,
    date: PropTypes.string
  }).isRequired
}

export default HistoryItem
