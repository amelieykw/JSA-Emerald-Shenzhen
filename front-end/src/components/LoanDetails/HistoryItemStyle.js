import { StyleSheet } from 'react-native'

const style = StyleSheet.create({
  body: {
    flexDirection: 'row',
    justifyContent: 'flex-start',
    marginVertical: 5
  },
  circle: {
    height: 45,
    width: 45,
    borderRadius: 23,
    alignItems: 'center',
    justifyContent: 'center',
    marginRight: 20
  },
  up: {
    backgroundColor: '#FC7043'
  },
  down: {
    backgroundColor: '#72BB53'
  },
  icon: {
    color: 'white',
    fontSize: 30,
    fontWeight: 'bold'
  },
  date: {
    fontSize: 15,
    color: 'grey'
  }
})

export default style
