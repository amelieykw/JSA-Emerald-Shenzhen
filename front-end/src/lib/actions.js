import URL from './URL'

const ACTION_TYPE = {
  LOADING: 'LOADING',
  LOADDONE: 'LOADDONE',
  INIT_LOANLIST: 'INIT_LOANLIST',
  INIT_LOANLIST_SUCCESS: 'INIT_LOANLIST_SUCCESS',
  INIT_PRODUCTLIST: 'INIT_PRODUCTLIST',
  INIT_PRODUCTLIST_SUCCESS: 'INIT_PRODUCTLIST_SUCCESS'
}

async function fetchData(url) {
  console.log(URL)
  console.log(`Outside fetchData return : ${Object.entries(URL)}`)
  console.log(`Outside fetchData return : ${URL.getProducts}`)
  return fetch(url, {
    headers: { Authentication: 'Bearer token', Accept: 'application/json' }
  }).then(res => res.json())
}

function fetchLoanList() {
  return async dispatch => {
    dispatch({ type: ACTION_TYPE.LOADING })
    dispatch({ type: ACTION_TYPE.INIT_LOANLIST })

    try {
      const data = await fetchData(URL.getLoans)
      dispatch({ type: ACTION_TYPE.INIT_LOANLIST_SUCCESS, loanList: data })
      dispatch({ type: ACTION_TYPE.LOADDONE })
    } catch (error) {
      // dispatch({ type: ACTION_TYPE.ERROR, error })
      console.log(error)
    }
  }
}

function fetchProductList() {
  console.log(`Outside fetchProductList try : ${Object.entries(URL)}`)
  console.log(`Outside fetchProductList try : ${URL.getProducts}`)
  return async dispatch => {
    dispatch({ type: ACTION_TYPE.LOADING })
    dispatch({ type: ACTION_TYPE.INIT_PRODUCTLIST })

    try {
      console.log(`Inside fetchProductList try : ${Object.entries(URL)}`)
      console.log(`Inside fetchProductList try : ${URL.getProducts}`)
      const data = await fetchData(URL.getProducts)
      dispatch({ type: ACTION_TYPE.INIT_PRODUCTLIST_SUCCESS, productList: data })
      dispatch({ type: ACTION_TYPE.LOADDONE })
    } catch (error) {
      // dispatch({ type: ACTION_TYPE.ERROR, error })
      console.log(error)
    }
  }
}

export { fetchLoanList, fetchProductList, ACTION_TYPE }
