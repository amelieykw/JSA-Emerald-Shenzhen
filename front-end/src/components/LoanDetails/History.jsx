import React from 'react'
import { View, ScrollView, FlatList, Text } from 'react-native'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'

import HistoryItem from './HistoryItem'
import style from './HistoryStyle'

const History = ({ loanHistory }) => {
  return (
    <View style={style.body}>
      {loanHistory ? (
        <>
          <Text style={style.title}>History</Text>
          <ScrollView style={{ height: 350 }}>
            <FlatList
              data={loanHistory}
              renderItem={({ item }) => <HistoryItem historyItem={item} />}
              keyExtractor={(_, index) => index.toString()}
            />
          </ScrollView>
        </>
      ) : (
        <View style={style.warning}>
          <Text style={style.warningText}>No available records now</Text>
        </View>
      )}
    </View>
  )
}

History.propTypes = {
  loanHistory: PropTypes.arrayOf(PropTypes.any)
}

History.defaultProps = {
  loanHistory: undefined
}

const mapStateToProps = state => {
  return {
    loanHistory: state.loanList.find(loanItem => loanItem.id === state.detailLoanId).history
  }
}

export default connect(mapStateToProps, null)(History)
