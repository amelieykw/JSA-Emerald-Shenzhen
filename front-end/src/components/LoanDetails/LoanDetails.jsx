import React, { useEffect } from 'react'
import { Content, Container, Spinner } from 'native-base'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'

import Detail from './Detail'
import Chart from './Chart'
import History from './History'

import { fetchDetail } from '../../lib/actions'

function LoanDetails({ loading, fetchData, id }) {
  useEffect(() => {
    fetchData(id)
  }, [])

  return (
    <Container>
      <Content>
        <Detail />
        {loading === 'DETAIL' ? (
          <Spinner color="blue" />
        ) : (
          <>
            <Chart />
            <History />
          </>
        )}
      </Content>
    </Container>
  )
}

const mapStateToProps = state => {
  return { loading: state.loading, id: state.detailLoanId }
}
const mapDispatchToProps = dispatch => {
  return {
    fetchData: id => dispatch(fetchDetail(id))
  }
}

LoanDetails.navigationOptions = {
  title: 'Loan Details'
}

LoanDetails.propTypes = {
  id: PropTypes.number.isRequired,
  fetchData: PropTypes.func.isRequired,
  loading: PropTypes.string.isRequired
}

export default connect(mapStateToProps, mapDispatchToProps)(LoanDetails)
