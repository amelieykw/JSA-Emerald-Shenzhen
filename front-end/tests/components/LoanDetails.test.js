import React from 'react'
import thunk from 'redux-thunk'
import { Provider } from 'react-redux'
import renderer from 'react-test-renderer'
import configureStore from 'redux-mock-store'

import data from '../../helpers/mockData_FE'
import { fetchDetail, ACTION_TYPE } from '../../src/lib/actions'
import LoanDetails from '../../src/components/LoanDetails/LoanDetails'

const mockStore = configureStore([thunk])

describe('<LoanDetails />', () => {
  let component
  let payloads
  const store = mockStore({ detailLoanId: 1, loanList: data.detailedLoans, loading: true })
  beforeEach(() => {
    payloads = [
      {
        type: ACTION_TYPE.LOADING
      },
      {
        type: ACTION_TYPE.INIT_DETAIL
      },
      {
        type: ACTION_TYPE.INIT_DETAIL_SUCCESS,
        loanList: []
      },
      {
        type: ACTION_TYPE.LOADDONE
      }
    ]
  })

  const mockGetParam = jest.fn()
  mockGetParam.mockReturnValue({ id: 1, type: { name: 'mockType' }, remaining: 200 })

  it('<LoanDetails /> should render correctly', () => {
    component = renderer.create(
      <Provider store={store}>
        <LoanDetails />
      </Provider>
    )
    expect(component.toJSON()).toMatchSnapshot()
  })

  it('<LoanDetails /> should dispatch an action when mounted', async () => {
    renderer.act(() => {
      component = renderer.create(
        <Provider store={store}>
          <LoanDetails />
        </Provider>
      )
    })

    // eslint-disable-next-line no-undef
    window.fetch = jest.fn().mockImplementation(() =>
      Promise.resolve({
        json: () => Promise.resolve({})
      })
    )
    await store.dispatch(fetchDetail())
    const actions = store.getActions()

    expect(actions.pop()).toStrictEqual(payloads.pop())
  })
})
