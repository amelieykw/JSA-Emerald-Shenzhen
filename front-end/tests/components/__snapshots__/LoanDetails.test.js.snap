// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoanDetails /> <LoanDetails /> should render correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#fff",
      "flex": 1,
      "height": 1334,
    }
  }
>
  <RCTScrollView
    automaticallyAdjustContentInsets={false}
    contentContainerStyle={
      Array [
        Object {
          "padding": undefined,
        },
        undefined,
      ]
    }
    contentInset={
      Object {
        "bottom": 0,
      }
    }
    enableAutomaticScroll={true}
    enableOnAndroid={false}
    enableResetScrollToCoords={true}
    extraHeight={75}
    extraScrollHeight={0}
    getScrollResponder={[Function]}
    handleOnScroll={[Function]}
    keyboardDismissMode="interactive"
    keyboardOpeningTime={250}
    keyboardShouldPersistTaps="handled"
    keyboardSpace={0}
    onScroll={[Function]}
    resetKeyboardSpace={[Function]}
    resetScrollToCoords={
      Object {
        "x": 0,
        "y": 0,
      }
    }
    scrollEventThrottle={1}
    scrollForExtraHeightOnAndroid={[Function]}
    scrollIntoView={[Function]}
    scrollToEnd={[Function]}
    scrollToFocusedInput={[Function]}
    scrollToPosition={[Function]}
    showsVerticalScrollIndicator={true}
    style={
      Object {
        "backgroundColor": "transparent",
        "flex": 1,
      }
    }
    update={[Function]}
    viewIsInsideTabBar={false}
  >
    <View>
      <View
        style={
          Object {
            "backgroundColor": "#E3F2FD",
            "flex": 1,
            "flexDirection": "row",
            "height": 120,
            "padding": 15,
          }
        }
      >
        <View
          style={
            Object {
              "flex": 1,
              "flexDirection": "column",
              "justifyContent": "space-between",
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#000",
                  "fontFamily": "System",
                  "fontSize": 16,
                },
                Object {
                  "color": "#7A7A7A",
                  "fontSize": 18,
                },
              ]
            }
            uppercase={false}
          >
            ID - 1
          </Text>
          <Text
            style={
              Array [
                Object {
                  "color": "#000",
                  "fontFamily": "System",
                  "fontSize": 16,
                },
                Object {
                  "color": "#5A5C60",
                  "fontSize": 32,
                },
              ]
            }
            uppercase={false}
          >
            Student Loan
          </Text>
          <Text
            style={
              Object {
                "color": "#000",
                "fontFamily": "System",
                "fontSize": 16,
              }
            }
            uppercase={false}
          >
            Current Interest rate: 0.03%
          </Text>
        </View>
        <View
          style={
            Object {
              "flex": 1,
              "flexDirection": "column",
              "justifyContent": "space-between",
            }
          }
        >
          <View>
            <Text
              style={
                Array [
                  Object {
                    "color": "#000",
                    "fontFamily": "System",
                    "fontSize": 16,
                  },
                  Object {
                    "alignSelf": "flex-end",
                    "color": "#FC815E",
                    "fontSize": 36,
                  },
                ]
              }
              uppercase={false}
            >
              $ 3400
            </Text>
            <Text
              style={
                Array [
                  Object {
                    "color": "#000",
                    "fontFamily": "System",
                    "fontSize": 16,
                  },
                  Object {
                    "alignSelf": "flex-end",
                    "color": "#7A7A7A",
                    "fontSize": 12,
                    "paddingRight": 5,
                  },
                ]
              }
              uppercase={false}
            >
              remaining
            </Text>
          </View>
          <View
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "flexDirection": "row",
                "justifyContent": "flex-end",
                "opacity": 1,
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "black",
                  "borderRadius": 20,
                  "height": 21,
                  "marginRight": 8,
                  "width": 21,
                }
              }
            >
              <Text />
            </View>
            <Text
              style={
                Object {
                  "color": "#000",
                  "fontFamily": "System",
                  "fontSize": 16,
                }
              }
              uppercase={false}
            >
              PAY NOW
            </Text>
          </View>
        </View>
      </View>
      <View
        style={
          Object {
            "paddingHorizontal": 30,
            "paddingVertical": 10,
          }
        }
      >
        <Text
          style={
            Object {
              "fontSize": 20,
              "marginBottom": 20,
            }
          }
        >
          History
        </Text>
        <RCTScrollView
          data={
            Array [
              Object {
                "amount": -900,
                "date": "2019-11-22",
                "remaining": 3400,
                "text": "$900 payment received",
                "type": "payment",
              },
              Object {
                "amount": 125,
                "date": "2019-01-03",
                "remaining": 4300,
                "text": "2018 yearly interest $125",
                "type": "interest",
              },
            ]
          }
          disableVirtualization={false}
          getItem={[Function]}
          getItemCount={[Function]}
          horizontal={false}
          initialNumToRender={10}
          keyExtractor={[Function]}
          maxToRenderPerBatch={10}
          numColumns={1}
          onContentSizeChange={[Function]}
          onEndReachedThreshold={2}
          onLayout={[Function]}
          onMomentumScrollEnd={[Function]}
          onScroll={[Function]}
          onScrollBeginDrag={[Function]}
          onScrollEndDrag={[Function]}
          removeClippedSubviews={false}
          renderItem={[Function]}
          scrollEventThrottle={50}
          stickyHeaderIndices={Array []}
          updateCellsBatchingPeriod={50}
          viewabilityConfigCallbackPairs={Array []}
          windowSize={21}
        >
          <View>
            <View
              onLayout={[Function]}
              style={null}
            >
              <View
                style={
                  Object {
                    "flexDirection": "row",
                    "justifyContent": "flex-start",
                    "marginVertical": 5,
                  }
                }
              >
                <View
                  style={
                    Array [
                      Object {
                        "alignItems": "center",
                        "borderRadius": 23,
                        "height": 45,
                        "justifyContent": "center",
                        "marginRight": 20,
                        "width": 45,
                      },
                      Object {
                        "backgroundColor": "#72BB53",
                      },
                    ]
                  }
                >
                  <Text />
                </View>
                <View>
                  <Text
                    style={
                      Object {
                        "color": "#000",
                        "fontFamily": "System",
                        "fontSize": 16,
                      }
                    }
                    uppercase={false}
                  >
                    $900 payment received
                  </Text>
                  <Text
                    style={
                      Array [
                        Object {
                          "color": "#000",
                          "fontFamily": "System",
                          "fontSize": 16,
                        },
                        Object {
                          "color": "grey",
                          "fontSize": 15,
                        },
                      ]
                    }
                    uppercase={false}
                  >
                    2019-11-22
                  </Text>
                </View>
              </View>
            </View>
            <View
              onLayout={[Function]}
              style={null}
            >
              <View
                style={
                  Object {
                    "flexDirection": "row",
                    "justifyContent": "flex-start",
                    "marginVertical": 5,
                  }
                }
              >
                <View
                  style={
                    Array [
                      Object {
                        "alignItems": "center",
                        "borderRadius": 23,
                        "height": 45,
                        "justifyContent": "center",
                        "marginRight": 20,
                        "width": 45,
                      },
                      Object {
                        "backgroundColor": "#FC7043",
                      },
                    ]
                  }
                >
                  <Text />
                </View>
                <View>
                  <Text
                    style={
                      Object {
                        "color": "#000",
                        "fontFamily": "System",
                        "fontSize": 16,
                      }
                    }
                    uppercase={false}
                  >
                    2018 yearly interest $125
                  </Text>
                  <Text
                    style={
                      Array [
                        Object {
                          "color": "#000",
                          "fontFamily": "System",
                          "fontSize": 16,
                        },
                        Object {
                          "color": "grey",
                          "fontSize": 15,
                        },
                      ]
                    }
                    uppercase={false}
                  >
                    2019-01-03
                  </Text>
                </View>
              </View>
            </View>
          </View>
        </RCTScrollView>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
